:tada:  CI secrets: migration to GSM

antixar opened the pull request:
## What
Details [here](https://github.com/airbytehq/airbyte/issues/3732)

## How
This decision provides backward compatibility with all secrets workflows. All old settings should work after this changes.

### How to add permissions for working with Google Secret Manager(GSM):
Each developer should have the role `Development_CI_Secrets` of the project `dataline-integration-testing`
Now these permissions are provided for GL Python/Java teams only.

### How to add **new** secrets:
1. Go to your developer cloud [console](https://console.cloud.google.com/security/secret-manager?referrer=search&orgonly=true&project=dataline-integration-testing&supportedpurview=organizationId)
2. Create a new secret with the following rules:
  * Secret names can only contain English capital letters (A-Z), numbers (0-9), dashes (-), and underscores (_) 
  * It must be started with the prefix: `SECRET_`
3. Add labels:
    - connector (required) -- unique connector's name or set of connectors' names with '_' as delimiter
    - filename (optional) -- custom target secret file. Unfortunately Google doesn't use '.' into labels' values and so Airbyte CI scripts will add '.json' to the end automatically. By default secrets will be saved to `./secrets/config.json `
 4. Run the `test` command  for your connector/
 
For examples:
1. SECRET_SOURCE-S3_CREDS
    Labels:
    - connector=source-s3
    Result:  value => `airbyte-integrations/connectors/source-s3/secrets/config.json`
2. SECRET_SOURCE-ZENDESK_SUPPORT_AUTH_CREDS
    Labels:
    - connector=source-zendesk-support
    - filename=config_auth
    Result:  value => `airbyte-integrations/connectors/source-zendesk-support/secrets/config_auth.json`
3.  SECRET_GOOGLE_CREDS
    Labels:
    - connector=source-google-analytics_source-google-drive
    Result:  
    - value => `airbyte-integrations/connectors/source-google-analytics/secrets/config.json` 
    - value => `airbyte-integrations/connectors/source-google-drive/secrets/config.json` 
    
### How to migrate to the new secrets' logic:
1. Create all necessary secrets how it is explained above.
2. Remove all lines with old GutHub secrets into this file: `tools/bin/ci_credentials.sh`
3. Remove all old secrets from Github

## Recommended reading order
1. `.github/workflows/{publish|test}-command.yml` - removed all secrets' values because the new script version will select necessary ones only.
2. `tools/lib/gcp-token.sh` - functions for generation of Google tokens. It can be used not only for the current task.
3.  `tools/bin/ci_credentials.sh` - changes of all CI secrets' logic

## Pre-merge Checklist

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-zendesk-support

antixar commented:
/test connector=connectors/source-s3

sherifnada commented on .github/workflows/test-command.yml:
```diff
@@ -65,131 +65,10 @@ jobs:
+        run: ./tools/bin/ci_credentials.sh test ${{ github.event.inputs.connector }}
         env:
-          AMAZON_SELLER_PARTNER_TEST_CREDS: ${{ secrets.AMAZON_SELLER_PARTNER_TEST_CREDS }}
-          AMAZON_ADS_TEST_CREDS: ${{ secrets.AMAZON_ADS_TEST_CREDS }}
```
did you create all the new secrets in GSM?

antixar replied:
No, it's not needed. I think all developers will migrate connectors' secrets during their new  enhancements and bugfixes gradually. GSM secrets are more prioritized for CI but if some new GSM secret is skipped, the CI script will use old github values.
Certainly we can copy all values to GSM but I'm not sure about their relevance because their source of truth is LastPass where now there is a complete mess

sherifnada replied:
No need to do it as part of this PR

sherifnada commented on .github/workflows/publish-command.yml:
```diff
@@ -70,131 +70,10 @@ jobs:
-          SOURCE_FRESHSERVICE_TEST_CREDS: ${{ secrets.SOURCE_FRESHSERVICE_TEST_CREDS }}
-          SOURCE_LEMLIST_TEST_CREDS: ${{ secrets.SOURCE_LEMLIST_TEST_CREDS }}
-          SOURCE_STRAVA_TEST_CREDS: ${{ secrets.SOURCE_STRAVA_TEST_CREDS }}
+          SECRETS_JSON: ${{ toJson(secrets) }}
```
could you add am ore descriptive name? maybe `GITHUB_PROVIDED_SECRETS_JSON`?

antixar replied:
Done

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+      --header "authorization: Bearer ${access_token}" \
+      --header "content-type: application/json" \
+      --header "x-goog-user-project: ${project_id}")
+    [[ -z ${data} ]] && error "Can't load secrets' list"
```
would this help in debugging failures? 
```suggestion
    [[ -z ${data} ]] && error "Can't load secret for connector ${CONNECTOR_NAME}"
```

antixar replied:
Done

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+      --header "content-type: application/json" \
+      --header "x-goog-user-project: ${project_id}")
+    [[ -z ${data} ]] && error "Can't load secrets' list"
+    [[ ${data} == "{}" ]] && data='{"secrets": []}'
```
can you add a comment mentioning why this is needed?

antixar replied:
Added

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+        echo "Not found ${CONNECTOR_NAME} info into the label 'connector' of the secret ${secret_name}"
+        continue
+      fi
+      # skip secrets for other comments
```
```suggestion
      # skip secrets for other commands
```

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+      # all secrets without the "command" label will be added too
+      [[ ${label_command} != ${COMMAND_NAME} ]] && continue
+      # all secret file names should be finished with ".json"
+      # but '.' cant be used
```
```suggestion
      # but '.' cant be used in google, so we append it
```

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
   local connector_name=$1
   local creds=$2
   local cred_filename=${3:-config.json}
+  local source_name=${4:-github}
```
I suggest renaming this to `secrets_provider_name` to avoid overloading with connector sources

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+}
+
+
+function export_github_secrets(){
```
We should also assert this var is available
```suggestion
function export_github_secrets(){
  # We expect that all secrets injected from github are available in an env variable `SECRETS_JSON`
```

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
 
 set -e
 
+COMMAND_NAME=$1
```
```suggestion
# This script loads secrets from Google Secrets Manager into the relevant connector's working directory
# If a secret is available in both Github and GSM, then the GSM secret is used, otherwise Github. 
COMMAND_NAME=$1
```

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+COMMAND_NAME=$1
+
+# all secrets will be loaded if the second argument is not present
+CONNECTOR_FULLNAME=${2:-all}
```
should we move this after all the method declarations to cetnralize where all the logic for the `main` method exists?

antixar replied:
all bash function arguments have separate namespaces ($1, $2 ....) Thus script's arguments can be available in its "global" namespace only. It is a const variable and I defined it at the begging file.

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+
+  # docs: https://cloud.google.com/secret-manager/docs/filtering#api
+  local filter="name:SECRET_"
+  [[ ${CONNECTOR_NAME} != "all" ]] && filter="${filter} AND labels.connector:${CONNECTOR_NAME}"
```
i think this is what's happening but just to confirm: we are not loading/listing all secrets from GSM; only the ones relevant to the current connector, correct?

antixar replied:
Yes, correctly.
in general using of labels are enough but I want to stimulate developers that they will use the common naming convention.

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+declare -A SECRET_MAP
+
+
 function write_standard_creds() {
```
suggest renaming to `read_secrets`

antixar replied:
@sherifnada,
this is the single debatable function name. this function is used for the old logic (that huge mapping list for secrets).
Sure, we can rename it but contributors will have some concern if they don't read updated secrets' flow docs before.
Thus I saved its old name.

sherifnada replied:
I think devs will be able to pick up on what's happening here; renaming will pose a problem. It's better that we have a slight confusion that accurately reflects the current/future state rather than stay beholden to decisions from the past :)

antixar replied:
ok, but I will rename it immediately before merging to decrease conflicts' amount

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+  return 0
+}
+
+function _write_standard_creds() {
```
we can probably lose the `standard` terminology and rename to `write_secret_to_disk`

antixar replied:
Done

sherifnada commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
   echo "$creds" > "${secrets_dir}/${cred_filename}"
 }
 
+function save_all() {
```
suggest renaming to `write_all_secrets`

antixar replied:
Done

sherifnada reviewed (commented):
This looks great! Some small comments here and there but overall the approach looks really solid and the instructions are very helpful. 

1. Thanks a lot for the PR description. it is super helpful. Before merging this PR, we should document it in a markdown document, possibly under the Connector Development section. It probably deserves its own page. Also let's update the PR template with a link to these instructions. 
2. Why did you add the feature to separate credentials used in `/test` and `/publish`? It seems like a potentially useful feature, but it feels like it's potentially enabling an anti-pattern i.e: the credentials used in testing should be the same used in publishing. Otherwise running `/publish` and `/test` will be running tests against different connectors.

tuliren commented on tools/bin/ci_credentials.sh:
```diff
@@ -1,26 +1,150 @@
+    return 0
+  fi
+  local key="${connector_name}#${cred_filename}"
+  [[ -z "${creds}" ]] && error "Don't find data for the connector '${key})"
```
```suggestion
  [[ -z "${creds}" ]] && error "Empty credential for the connector '${key})"
```

tuliren reviewed (approved):
Looks good!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=7528) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=7528) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: antixar<br/>:x: Maksym Pavlenok<br/><hr/>**Maksym Pavlenok** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=7528) it.</sub>

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-zendesk-support

antixar commented:
@sherifnada,
About your points:
 1. I've updated the official CI docfile. Please check
 2. I was not sure to both CI secrets' suites are same for `test` and `publish` commands. Thanks that you threw away my concern. I've removed this logic from my updates.

antixar commented:
/test connector=connectors/source-zendesk-support

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-zendesk-support

sherifnada commented on docs/connector-development/README.md:
```diff
@@ -130,10 +130,15 @@ Once you've finished iterating on the changes to a connector as specified in its
+
+#### How to migrate to the new secrets' logic:
+1. Create all necessary secrets how it is explained above.
+2. Remove all lines with old connector's GutHub secrets into this file: tools/bin/ci_credentials.sh
```
```suggestion
2. Remove all lines with old connector's Github secrets from this file: tools/bin/ci_credentials.sh
```

sherifnada commented on docs/connector-development/README.md:
```diff
@@ -130,10 +130,15 @@ Once you've finished iterating on the changes to a connector as specified in its
+#### How to migrate to the new secrets' logic:
+1. Create all necessary secrets how it is explained above.
+2. Remove all lines with old connector's GutHub secrets into this file: tools/bin/ci_credentials.sh
+3. Remove all old secrets from Githubthe secret from env variables to the connector directory**: edit `tools/bin/ci_credentials.sh` to write the secret into the `secrets/` directory of the relevant connector.  
```
Should this be written as follows? 
```suggestion
3. Remove all old secrets from Github repository secrets.
```

sherifnada commented on docs/connector-development/README.md:
```diff
@@ -130,10 +130,15 @@ Once you've finished iterating on the changes to a connector as specified in its
-2. **Add the credentials to Github Secrets**: To inject credentials into a CI workflow, the first step is to add it to Github Secrets, specifically within the ["more-secrets" environment](https://github.com/airbytehq/airbyte/settings/environments/276695501/edit). Admin access to the Airbyte repo is required to do this. All Airbyte engineers have admin access and should be able to do this themselv ...
-3. **Inject the credentials into test and publish CI workflows**: edit the files `.github/workflows/publish-command.yml` and `.github/workflows/test-command.yml` to inject the secret into the CI run. This will make these secrets available to the `/test` and `/publish` commands.
-4. **During CI, write the secret from env variables to the connector directory**: edit `tools/bin/ci_credentials.sh` to write the secret into the `secrets/` directory of the relevant connector.  
+1. **Place the credentials into Google Secret Manager(GSM)**: Airbyte uses a project 'Google Secret Manager' service as the source of truth for all secrets. Place the credentials **exactly as they should be used by the connector** into a GSM secret i.e.: it should basically be a copy paste of the `config.json` passed into a connector via the `--config` flag. We use the following naming pattern: ` ...
```
Could you also add instructions about how to add contributors to this GCP project? specifically, which permissions do they need to have?

antixar replied:
Added [here](https://github.com/airbytehq/airbyte/pull/7528/files?authenticity_token=nwxI%2BzEDpzJQEIYJrQ8w%2FczTB0UlxPoTiTPxTbDrE4aCmrl9U7PvDrESwJo%2BHo%2BOy6sCA7BUNpToXqRIlTXCXA%3D%3D&file-filters%5B%5D=.md&file-filters%5B%5D=.sh#diff-c4afbe4289e28951ba8fd39b072c5d19a0a4ad4d148f38f9a4fa35cc9bc81381R139)

sherifnada commented on docs/connector-development/README.md:
```diff
@@ -130,10 +130,15 @@ Once you've finished iterating on the changes to a connector as specified in its
-2. **Add the credentials to Github Secrets**: To inject credentials into a CI workflow, the first step is to add it to Github Secrets, specifically within the ["more-secrets" environment](https://github.com/airbytehq/airbyte/settings/environments/276695501/edit). Admin access to the Airbyte repo is required to do this. All Airbyte engineers have admin access and should be able to do this themselv ...
-3. **Inject the credentials into test and publish CI workflows**: edit the files `.github/workflows/publish-command.yml` and `.github/workflows/test-command.yml` to inject the secret into the CI run. This will make these secrets available to the `/test` and `/publish` commands.
-4. **During CI, write the secret from env variables to the connector directory**: edit `tools/bin/ci_credentials.sh` to write the secret into the `secrets/` directory of the relevant connector.  
+1. **Place the credentials into Google Secret Manager(GSM)**: Airbyte uses a project 'Google Secret Manager' service as the source of truth for all secrets. Place the credentials **exactly as they should be used by the connector** into a GSM secret i.e.: it should basically be a copy paste of the `config.json` passed into a connector via the `--config` flag. We use the following naming pattern: ` ...
```
we should probably make a script for handling this process e.g: `write-ci-secret-to-gsm <connector_name>` that handles this whole process. This can be a follow up issue prioritized later, doesn't need to be in this PR of course.

antixar replied:
@sherifnada ,
Sure, it is not problem to create an one-time script. I've created the [follow-up task](https://github.com/airbytehq/airbyte/issues/7774) for this.
And we need to wait some time after starting the GSM new logic ( +- 1 month). Perhaps contributors will detect some new issues or cases which are incompatible with it.

sherifnada reviewed (commented):
Can you let me know once you've written up .md instructions for how to add devs to the GCP project? I'll add everyone to that project

antixar commented:
@sherifnada ,
1. About access to GSM, we can use Google groups for this. I have already added 2 our ones:  
```
airbyte_python_team@globallogic.com
airbyte_java_team@globallogic.com
```
2. the function` write_standard_creds ` will be renamed immediately before merging

antixar commented:
/test connector=connectors/source-zendesk-support

sherifnada reviewed (approved):
very exciting!

antixar commented:
/test connector=connectors/source-zendesk-support

antixar commented:
/test connector=connectors/source-s3

antixar commented:
/test connector=connectors/source-zendesk-support

antixar commented:
/test connector=connectors/source-commercetools
